### ISSUE TITLE
[BUG] azure-messaging-eventhubs silently fails to read messages

### ISSUE THREAD

**Issue by schatba** [Azure/azure-sdk-for-java#37925]
**Describe the bug**
This bug is in response to Microsoft Support request 2311210050003302, to which we were advised to open a bug on the SDK.

We have multiple instances of a service reading from an eventhub and observed one instance having vastly increased demand for memory. We suspect there were malformed messages which are silently being dismissed by the Azure SDK but read over and over, causing the high outgoing load on the eventhub and the huge memory load on the one instance that processed this partition. We did however, not see any errors or exceptions in our logs that would explain any of this behaviour.

Every processed message is checkpointed, so no double reads occur. Log level is set to INFO.
We know that logging works as expected since we regularly get the exceptions of lost partition ownership when the partition rebalancing happens.

We 'solved' the problem by purging the - what we speculate to be - problematic messages by temporarily decreasing the retentiontime on the eventhub to 1h. However, were this to occur in our production environment, this 'solution' would cause data loss for our customers.
Luckily this problem occurred on our staging environment, however we need an urgent fix to avoid the same problem on production. Since there is no exception being thrown, we have no way to mitigate or even recognize the issue other than monitoring the eventhub directly.

We opened a Microsoft support ticket to have this investigated. The results are that there were a lot of client exceptions thrown on the eventhub (see screenshot below), which happened because the consumers were disconnected frequently in the time frame where the increased reads were observed. To reiterate, after the messages were purged and no changes were made to the service, operations continued smoothly. So the issue seems to be that the Azure SDK has problems reading certain messages from the eventhub, but again, there are no exceptions to indicate what might be wrong.

***Exception or Stack Trace***
Nothing being logged or thrown is the problem.

**To Reproduce**
- our service was not able to read these messages, so unfortunately the way to reproduce is unknown

**Expected behavior**
If there is a problem with reading certain messages, there needs to be exceptions thrown to indicate the issue.

**Screenshots**
![image](https://github.com/Azure/azure-sdk-for-java/assets/152964423/607eb95f-0b24-4b6d-8ccd-8a9fd24d00bf)

Massively increased Outgoing Messages on the eventhub (the drop-off and return to normal happened once we purged by reducing the retentiontime)
![image](https://github.com/Azure/azure-sdk-for-java/assets/152964423/47bae4e7-1305-43de-8a97-2b2046d16a10)

Memory usage on our pods (that's 16gb on the one which consumes the troublesome partition; also returned to normal after the purge):
![image](https://github.com/Azure/azure-sdk-for-java/assets/152964423/e5b78a78-5c1a-4fce-befd-41ca7e428c63)

**Setup (please complete the following information):**
 - OS: CBL-Mariner 2.0 Distroless
 - IDE: not relevant
 - Library/Libraries: 
   - `com.azure:azure-messaging-eventhubs:5.15.8`
   - `com.azure:azure-messaging-eventhubs-checkpointstore-blob:1.16.9`
   - `com.azure:azure-identity:1.10.0`
   - `com.azure.spring:azure-spring-cloud-appconfiguration-config:2.11.0`
   - `com.microsoft.azure:applicationinsights-web:3.4.14`
   - `com.microsoft.azure:applicationinsights-runtime-attach:3.4.14`
 - Java version: 17
 - App Server/Environment: Tomcat
 - Frameworks: Spring Boot 3.1.3

I do not suspect a dependency version mismatch.

**Code Snippets**
Not really relevant to the problem, but just in case, our eventhub consumer:
```java
 public void startTelemetryProcessor() {
    LOGGER.debug("starting new EventProcessorClient");
    String fullyQualifiedNamespace = config.getIncomingEventHubNamespace() + config.getEhHostSuffix();
    String consumerGroupName = config.getConsumerGroupName();
    TokenCredential tokenCredential = azureCredentials.getTokenCredential();

    BlobContainerAsyncClient blobContainerClient = createBlobClient(tokenCredential, config);

    telemetryProcessorHost = new EventProcessorClientBuilder()
            .fullyQualifiedNamespace(fullyQualifiedNamespace)
            .eventHubName(config.getIncomingEventHub())
            .credential(tokenCredential)
            .consumerGroup(consumerGroupName)
            .checkpointStore(new BlobCheckpointStore(blobContainerClient))
            .loadBalancingUpdateInterval(Duration.ofSeconds(5))
            .partitionOwnershipExpirationInterval(Duration.ofSeconds(20))
            .loadBalancingStrategy(LoadBalancingStrategy.GREEDY)
            .processEventBatch(createEventConsumer(), 150, Duration.ofSeconds(2))
            .processError(createErrorConsumer())
            .processPartitionInitialization(createInitializationConsumer())
            .processPartitionClose(createCloseConsumer())
            .buildEventProcessorClient();
    try {
      telemetryProcessorHost.start();
      LOGGER.info("Started EventProcessorClient for telemetry with identifier {}", telemetryProcessorHost.getIdentifier());
    } catch (Exception ex) {
      LOGGER.error("Failed at startup telemetry EventProcessorClient", ex);
    }
  }

  private Consumer<EventBatchContext> createEventConsumer() {
    return eventDataBatch -> {
      if (!eventDataBatch.getEvents().isEmpty()) {
        try {
          String transactionId = UUID.randomUUID().toString();
          TelemetryProcessResult result = eventService.processTelemetry(transactionId, eventDataBatch.getEvents());
          eventService.publishMessages(transactionId, result)
                  .thenCompose(stage -> eventDataBatch.updateCheckpointAsync().toFuture())
                  .thenRun(() -> LOGGER.debug("Checkpointed Partition {}", eventDataBatch.getPartitionContext().getPartitionId()))
                  .get();
        } catch (InterruptedException e) {
          LOGGER.error("Error while processing message", e);
          Thread.currentThread().interrupt();
        } catch (ExecutionException e) {
          LOGGER.error("Error while processing message", e);
        }
      }
    };
  }

  private Consumer<ErrorContext> createErrorConsumer() {
    return context -> {
      PartitionContext partitionContext = context.getPartitionContext();
      LOGGER.error("Error received for Partition {} on {}",
              partitionContext.getPartitionId(),
              partitionContext.getEventHubName(),
              context.getThrowable());
    };
  }

  private Consumer<CloseContext> createCloseConsumer() {
    return closeContext -> {
      metrics.decrementPartitionCount();
      PartitionContext partitionContext = closeContext.getPartitionContext();
      LOGGER.info("Closed partition {} on {}. Reason: {}",
              partitionContext.getPartitionId(),
              partitionContext.getEventHubName(),
              closeContext.getCloseReason());
    };
  }

  private Consumer<InitializationContext> createInitializationConsumer() {
    return initializationContext -> {
      metrics.incrementPartitionCount();
      PartitionContext partitionContext = initializationContext.getPartitionContext();
      LOGGER.info("Opened partition {} on {}",
              partitionContext.getPartitionId(),
              partitionContext.getEventHubName());
    };
  }
```

**Additional context**
Our service is running in a Docker container in an Azure kubernetes cluster, using a workload identity to authenticate with the eventhub.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by conniey** [comment]
[10:32 AM] Connie Yau
Hey,
Thanks for reporting this issue! I have a couple of questions and comments:
* azure-messaging-eventhubs 5.15.3 is from March.  Since then, we've done several [fixes in the underlying library](CHANGELOG.md link in azure-core-amqp).  Does this reproduce when you move to 5.17.0?
* What is the exception you're seeing during this burst phase?

> "Every processed message is checkpointed, so no double reads occur. Log level is set to INFO."
> The results are that there were a lot of client exceptions thrown on the eventhub (see screenshot below), which happened because the consumers were disconnected frequently in the time frame where the increased reads were observed. 

It is normal for there to be more outgoing messages than incoming when you are running multiple instances of the EventProcessorClient with LoadBalancingStrategy.GREEDY.  Whenever an EventProcessorClient claims an Event Hub partition, by default, it prefetches 500 messages.  There is a load-balancing cycle that happens in each running instance of EventProcessorClient where instances can steal partitions from other clients to balance the number of partitions being processed. In this case, there will be another 500 messages prefetched.

Example: 
* 2 clients (A and B).
* Event Hub has 3 partitions (0, 1, 2).
1. Client A starts and claims 3 partitions (0, 1, 2).
   1. Each receiver will prefetch 500 messages (assuming default prefetch). 1500 outgoing messages.
2. Client B starts a few seconds after, notices that the load is not balanced, so it will randomly steal a partition from Client A; partition 1.
   1. Client B's receiver for partition 1 will start at the last checkpointed place and prefetch 500 messages.
   2. When the receiver starts, Client A's receiver is kicked off of the partition.  The message looks like "[New receiver 'nil' with higher epoch of '0' ..."](https://learn.microsoft.com//azure/developer/java/sdk/troubleshooting-messaging-event-hubs-processor#current-receiver-receiver_name-with-epoch-0-is-getting-disconnected).

Client A may have not processed all 500 prefetched messages so there will be some overlapping EventData sent to Client B.  The instances are constantly refilling their prefetch queue as the client processes a message.

You can find more information in our troubleshooting guide: https://learn.microsoft.com/azure/developer/java/sdk/troubleshooting-messaging-event-hubs-processor#partition-ownership-changes-frequently

**Comment by schatba** [comment]
> azure-messaging-eventhubs 5.15.3 is from March. Since then, we've done several [fixes in the underlying library](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-amqp/CHANGELOG.md). Does this reproduce when you move to 5.17.0?

We'll upgrade, sure. But we won't be able to reproduce the problem. In hindsight we should've kept the troublesome messages for further analysis.

> It is normal for there to be more outgoing messages than incoming when you are running multiple instances of the EventProcessorClient with LoadBalancingStrategy.GREEDY. Whenever an EventProcessorClient claims an Event Hub partition, by default, it prefetches 500 messages.

Over the course of 3 days we saw **34 million** additional message reads (see screenshot). Whenever loadbalancing happens, we also get and record the AMQP exceptions due to partition ownership being lost. We did not see any spikes in exceptions in our Application Insights, so there is no way that rebalances caused these _disproportionately_ increased reads. That is exactly the problem though. From the application's point of view, there was no problem.

**Comment by conniey** [comment]
This looks to be resolved by: https://github.com/Azure/azure-sdk-for-java/issues/38368, which was released on February as 5.18.1. https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md#bugs-fixed-1
